lib/pbench/server/db/models/tracker.py

```diff
@@ -0,0 +1,546 @@
+                    "path" if not present.
+                "state": The initial state of the new dataset.
+
+        Returns:
+            Dataset: A new Dataset object initialized with the keyword
+            parameters. (See `__init__` for details.)
+        """
+        try:
+            dataset = Dataset(**kwargs)
+            dataset.add()
+        except Exception as e:
+            Dataset.logger.error("Failed create: {}", str(e))
+            return None
+        else:
+            return dataset
+
+    @staticmethod
+    def attach(**kwargs):
+        """
+        attach Attempt to fetch dataset for the controller and dataset name,
+        or using a specified file path (see __init__ and _render_path).
+
+        If the `state` kwarg is specified, the dataset will be created in that
+        state or, if it already exists, `attach` will attempt to advance the
+        dataset to that state.
+
+        Args:
+            kwargs (dict):
+                "path": A tarball file path from which the controller (host)
+                    name, the tarball dataset name (basename minus extension)
+                    or both will be derived. The `path` key will be removed
+                    from the dict after processing. (The SQLAlchemy super
+                    constructor would complain about it as its not a column.)
+                "controller": The controller name (hostname) of the dataset;
+                    this is retained if specified, or will be constructed
+                    from "path" if not present.
+                "name": The dataset name (file path basename minus ".tar.xz");
+                    this is retained if specified, or will be constructed from
+                    "path" if not present.
+                "state": The desired state to advance the dataset.
+
+        Raises:
+            DatasetSqlError: problem interacting with Database
+            DatasetNotFound: the specified dataset doesn't exist
+            DatasetTerminalStateViolation: dataset is in terminal state and
+                can't be advanced
+            DatasetBadStateTransition: dataset cannot be advanced to the
+                specified state
+
+        Returns:
+            Dataset: a dataset object in the desired state (if specified)
+        """
+        # Make sure we have controller and name from path
+        Dataset._render_path(kwargs)
+        controller = kwargs.get("controller")
+        name = kwargs.get("name")
+        try:
+            dataset = (
+                Database.db_session.query(Dataset)
+                .filter_by(controller=controller, name=name)
+                .first()
+            )
+        except SQLAlchemyError as e:
+            Dataset.logger.warning(
+                "Error attaching {}>{}: {}", controller, name, str(e)
+            )
+            raise DatasetSqlError(f"Error attaching {controller}:{name}: {e}")
+        else:
```

Comment:
The Python `try/else` is an interesting construct; it's arguably never "necessary", but it can help to make intent clear. In this case, it calls out the direct dependency on `dataset` and protects it should (for example) another `except` be added that doesn't raise.

Reply:
It is "necessary" in exactly the case that you describe:  when there is an `except` clause which doesn't exit the scope but whose case should not execute the contents of the `else` clause.  However, anyone adding an `except` clause, particularly one which which purports to "handle" the exception, needs to be cognizant of the effect of control continuing after the `try` block -- I would leave it up to them to add the `else` clause when/if it's actually neeeded.

Reply:
I can respect and appreciate that.  My principal concern is that, in a similar vein, I've come to regard indentation (particularly more than a couple of levels) with grave suspicion, and this use of the `else` clause promotes indentation where it could otherwise be avoided (e.g., in yesterday's deep dive, Peter showed us code with these `else` clauses nested _three_ deep, with other nestings inside them).  I don't think this use improves the code's readability, and it opens a subsequent debate about exactly what code should go inside the `else` clause and what code should appear after it...when, in cases like this one, there is no functional difference.